## Content:
### 01.Importing libraries
### 02. Importing Dataframes
### 03. What are Data Consistency Checks?
#### 3.1. Mixed-Type Data
#### 3.2. Missing Values
#### 3.3. Duplicates
##### 3.3.1. Finding Duplicates
##### 3.3.2. Addressing Duplicates
#### 3.4. Tidying Up and Exporting Changes
### 04.Task
#### Step 2
#### Step 3
#### Step 4
#### Step 5
#### Step 6
#### Step 7
#### Step 9












# 01.Importing libraries 


In [3]:
import pandas as pd
import numpy as np
import os

# 02. Importing Dataframes

In [5]:
path= r'C:\Users\Bilel\Desktop\data_analyst_Carrer Foundry\Data Immersion\07-12-24 Instacart Basket Analysis'
df_prods = pd.read_csv(os.path.join(path,'02 Data', 'Original Data', 'products.csv'),index_col = False)
df_ords  = pd.read_csv(os.path.join(path,'02 Data', 'Prepared Data', 'orders_wrangled.csv'),index_col = False)


# 03. What are Data Consistency Checks?

In [7]:
df_ords.head()

Unnamed: 0  order_id  user_id  order_number  orders_day_of_week  \
0           0   2539329        1             1                   2   
1           1   2398795        1             2                   3   
2           2    473747        1             3                   3   
3           3   2254736        1             4                   4   
4           4    431534        1             5                   4   

   order_hour_of_day  days_since_prior_order  
0                  8                     NaN  
1                  7                    15.0  
2                 12                    21.0  
3                  7                    29.0  
4                 15                    28.0

In [8]:
df_ords = df_ords.drop (columns = ['Unnamed: 0'])
df_ords.head()

order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0   2539329        1             1                   2                  8   
1   2398795        1             2                   3                  7   
2    473747        1             3                   3                 12   
3   2254736        1             4                   4                  7   
4    431534        1             5                   4                 15   

   days_since_prior_order  
0                     NaN  
1                    15.0  
2                    21.0  
3                    29.0  
4                    28.0

In [9]:
df_ords.describe()

order_id       user_id  order_number  orders_day_of_week  \
count  3.421083e+06  3.421083e+06  3.421083e+06        3.421083e+06   
mean   1.710542e+06  1.029782e+05  1.715486e+01        2.776219e+00   
std    9.875817e+05  5.953372e+04  1.773316e+01        2.046829e+00   
min    1.000000e+00  1.000000e+00  1.000000e+00        0.000000e+00   
25%    8.552715e+05  5.139400e+04  5.000000e+00        1.000000e+00   
50%    1.710542e+06  1.026890e+05  1.100000e+01        3.000000e+00   
75%    2.565812e+06  1.543850e+05  2.300000e+01        5.000000e+00   
max    3.421083e+06  2.062090e+05  1.000000e+02        6.000000e+00   

       order_hour_of_day  days_since_prior_order  
count       3.421083e+06            3.214874e+06  
mean        1.345202e+01            1.111484e+01  
std         4.226088e+00            9.206737e+00  
min         0.000000e+00            0.000000e+00  
25%         1.000000e+01            4.000000e+00  
50%         1.300000e+01            7.000000e+00  
75%         1.600000e+01            1.500000e+01  
max         2.300000e+01            3.000000e+01

## 3.1. Mixed-Type Data


In [11]:
#Create a dataframe
df_test = pd.DataFrame()

In [12]:
# Crate a mixed type column
df_test['mix']= ['a','b',1,True]

In [13]:
df_test.head()

mix
0     a
1     b
2     1
3  True

In [14]:
# check for mixed types
#for col in df_test.columns.tolist():
 # weird = (df_test[[col]].applymap(type) != df_test[[col]].iloc[0].apply(type)).any(axis = 1)
  #if len (df_test[weird]) > 0:
   # print (col)

In [15]:
# correction of mixed data
# df_test['mix'] = df_test['mix'].astype('str') or 'int64' instead of 'str'

## 3.2. Missing Values

In [17]:
df_prods.isnull().sum()

product_id        0
product_name     16
aisle_id          0
department_id     0
prices            0
dtype: int64

In [18]:
df_nan= df_prods[df_prods['product_name'].isnull()==True]

In [19]:
df_nan

product_id product_name  aisle_id  department_id  prices
33             34          NaN       121             14    12.2
68             69          NaN        26              7    11.8
115           116          NaN        93              3    10.8
261           262          NaN       110             13    12.1
525           525          NaN       109             11     1.2
1511         1511          NaN        84             16    14.3
1780         1780          NaN       126             11    12.3
2240         2240          NaN        52              1    14.2
2586         2586          NaN       104             13    12.4
3159         3159          NaN       126             11    13.1
3230         3230          NaN       120             16    14.4
3736         3736          NaN        41              8    14.8
4283         4283          NaN        77              7    14.4
4790         4790          NaN        91             16    14.5
38187       38183          NaN        39             12    20.9
40444       40440          NaN       120             16    14.8

In [20]:
df_prods.shape

(49693, 5)

In [21]:
df_prods_clean = df_prods[df_prods['product_name'].isnull() ==False]


In [22]:
df_prods_clean.shape

(49677, 5)

In [23]:
#drop all missing values
#df_prods.dropna(inplace = True)

In [24]:
#drop only the NaNs from a particular column
#df_prods.dropna(subset = [‘product_name’], inplace = True)


## 3.3. Duplicates

#### 3.3.1. Finding Duplicates

In [27]:
df_dups=df_prods_clean[df_prods_clean.duplicated()]

In [28]:
df_dups

product_id                                       product_name  \
462           462                  Fiber 4g Gummy Dietary Supplement   
18459       18458                                         Ranger IPA   
26810       26808               Black House Coffee Roasty Stout Beer   
35309       35306  Gluten Free Organic Peanut Butter & Chocolate ...   
35495       35491                            Adore Forever Body Wash   

       aisle_id  department_id  prices  
462          70             11     4.8  
18459        27              5     9.2  
26810        27              5    13.4  
35309       121             14     6.8  
35495       127             11     9.9

#### 3.3.2. Addressing Duplicates

In [30]:
df_prods_clean.shape

(49677, 5)

In [31]:
df_prods_clean_no_dups = df_prods_clean.drop_duplicates()

In [32]:
df_prods_clean_no_dups.shape

(49672, 5)

## 3.4. Tidying Up and Exporting Changes

In [34]:
# df_prods_clean_no_dups.to_csv(os.path.join(path, '02 Data', 'Prepared Data', 'products_checked.csv'))

# <span style="color:red"> 4.Task

### Step 2

In [37]:
df_ords.describe()

order_id       user_id  order_number  orders_day_of_week  \
count  3.421083e+06  3.421083e+06  3.421083e+06        3.421083e+06   
mean   1.710542e+06  1.029782e+05  1.715486e+01        2.776219e+00   
std    9.875817e+05  5.953372e+04  1.773316e+01        2.046829e+00   
min    1.000000e+00  1.000000e+00  1.000000e+00        0.000000e+00   
25%    8.552715e+05  5.139400e+04  5.000000e+00        1.000000e+00   
50%    1.710542e+06  1.026890e+05  1.100000e+01        3.000000e+00   
75%    2.565812e+06  1.543850e+05  2.300000e+01        5.000000e+00   
max    3.421083e+06  2.062090e+05  1.000000e+02        6.000000e+00   

       order_hour_of_day  days_since_prior_order  
count       3.421083e+06            3.214874e+06  
mean        1.345202e+01            1.111484e+01  
std         4.226088e+00            9.206737e+00  
min         0.000000e+00            0.000000e+00  
25%         1.000000e+01            4.000000e+00  
50%         1.300000e+01            7.000000e+00  
75%         1.600000e+01            1.500000e+01  
max         2.300000e+01            3.000000e+01

## <span style="color:green"> observation 
'order_number': Has a min value of 1 and a max value of 100, which may mean orders are capped at 100 items.
'order_day of the week': Has min value of 0 and max value of 6 meaning the seven numbers (0-6) represent days of the week.
'order_hour_of_day': Has a min value of 0 and max value of 23 meaning the twenty-four numbers (0-23) represent 24 hrs of the day.
'days_since prior order': Has a count less than all other rows inndicating missing values. It also has a measure of zero indicating that some orders could be places within the same day and the max value is 30, which may indicate that orders are capped at 30 days or 1 month.



### Step 3

In [40]:
for col in df_ords.columns.tolist():
  weird = (df_ords[[col]].map(type) != df_ords[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_ords[weird]) > 0:
    print (col)

## <span style="color:green"> observation 
There is no mixed-type data present in the df_ords dataframe

In [42]:
# Function to detect columns with mixed data types 
def detect_mixed_types(df):
    # Initialize an empty list to hold columns with mixed types
    mixed_cols = []
    # Iterate through each column in the dataframe
    for col in df.columns:
        # Check if the column has more than one unique data type
        if df[col].apply(type).nunique() > 1:
            # If yes, append the column name to the list
            mixed_cols.append(col)
# Call the function for dataframe df_ords 
mixed_type_columns = detect_mixed_types(df_ords)
# Print out the columns with mixed data types
print("Columns with mixed data types:", mixed_type_columns)

Columns with mixed data types: None


## <span style="color:green"> observation 
I found this function to look for mixed-type data using AI , is it also correct ?

## Step 4
There is no mixed-type data present in the df_ords dataframe

In [45]:
df_ords.dtypes

order_id                    int64
user_id                     int64
order_number                int64
orders_day_of_week          int64
order_hour_of_day           int64
days_since_prior_order    float64
dtype: object

## Step 5

In [47]:
# check for missing values in df_ords dataframe
df_ords.isnull().sum()

order_id                       0
user_id                        0
order_number                   0
orders_day_of_week             0
order_hour_of_day              0
days_since_prior_order    206209
dtype: int64

## <span style="color:green"> observation 
As suspected before in step 2 , 'days_since_prior_order' has 206209 missing values but as its name suggests, for the first order, there is no 'days_since_prior_order' count. The missing values are due to this effect.

## Step 6

In [50]:
#Create a subset for missing values
df_ords_nan = df_ords[df_ords['days_since_prior_order'].isnull() == True]
df_ords_nan

order_id  user_id  order_number  orders_day_of_week  \
0         2539329        1             1                   2   
11        2168274        2             1                   2   
26        1374495        3             1                   1   
39        3343014        4             1                   6   
45        2717275        5             1                   3   
...           ...      ...           ...                 ...   
3420930    969311   206205             1                   4   
3420934   3189322   206206             1                   3   
3421002   2166133   206207             1                   6   
3421019   2227043   206208             1                   1   
3421069   3154581   206209             1                   3   

         order_hour_of_day  days_since_prior_order  
0                        8                     NaN  
11                      11                     NaN  
26                      14                     NaN  
39                      11                     NaN  
45                      12                     NaN  
...                    ...                     ...  
3420930                 12                     NaN  
3420934                 18                     NaN  
3421002                 19                     NaN  
3421019                 15                     NaN  
3421069                 11                     NaN  

[206209 rows x 6 columns]

In [51]:
# Filter rows where order_number = 1 and days_since_prior_order is NaN
count_nan = df_ords_nan[(df_ords_nan['order_number'] == 1) & (df_ords_nan['days_since_prior_order'].isna())].shape[0]

# Display the result
print("Number of 'days_since_prior_order' = NaN when 'order_number' = 1:", count_nan)


Number of 'days_since_prior_order' = NaN when 'order_number' = 1: 206209


In [92]:
#address the missing values using an appropriate method
df_ords_clean = df_ords
df_ords_clean['new_customer'] = df_ords['days_since_prior_order'].isnull() == True
df_ords_clean

order_id  user_id  order_number  orders_day_of_week  \
0         2539329        1             1                   2   
1         2398795        1             2                   3   
2          473747        1             3                   3   
3         2254736        1             4                   4   
4          431534        1             5                   4   
...           ...      ...           ...                 ...   
3421078   2266710   206209            10                   5   
3421079   1854736   206209            11                   4   
3421080    626363   206209            12                   1   
3421081   2977660   206209            13                   1   
3421082    272231   206209            14                   6   

         order_hour_of_day  days_since_prior_order  new_customer  
0                        8                     NaN          True  
1                        7                    15.0         False  
2                       12                    21.0         False  
3                        7                    29.0         False  
4                       15                    28.0         False  
...                    ...                     ...           ...  
3421078                 18                    29.0         False  
3421079                 10                    30.0         False  
3421080                 12                    18.0         False  
3421081                 12                     7.0         False  
3421082                 14                    30.0         False  

[3421083 rows x 7 columns]

## <span style="color:green"> observation 
Using the result of the filter, I confirm that all the missing values for 'days_since_prior_order' are for the first order.
So There is no need to remove these missing values because they are quite valuable.


## Step 7

In [54]:
# Run a check for duplicate values in df_ords dataframe
df_ords_dup = df_ords[df_ords.duplicated()]
df_ords_dup

Empty DataFrame
Columns: [order_id, user_id, order_number, orders_day_of_week, order_hour_of_day, days_since_prior_order]
Index: []

## <span style="color:green"> observation 
The dataframe created to check duplicates has returned empty, meaning there are no duplicates values in df_ords dataframe.

## Step 9

In [94]:
# Exporting the final cleaned df_prods and df_ords dataframes as ".csv" files
df_prods_clean_no_dups.to_csv(os.path.join(path, '02 Data', 'Prepared Data', 'products_checked.csv'))

df_ords_clean.to_csv(os.path.join(path, '02 Data','Prepared Data', 'orders_checked.csv'))